// overpack/dist/MessageEmitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Action;
(function (Action) {
    Action["TriggerBuild"] = "TriggerBuild";
    Action["Error"] = "Error";
    Action["Warning"] = "Warning";
    Action["CreateTask"] = "CreateTask";
    Action["StartTask"] = "StartTask";
    Action["SucceedTask"] = "SucceedTask";
    Action["FailTask"] = "FailTask";
    Action["SkipTask"] = "SkipTask";
})(Action = exports.Action || (exports.Action = {}));
var MessageEmitter = (function () {
    function MessageEmitter(workerId, emit) {
        this.workerId = workerId;
        this.emit = emit;
    }
    MessageEmitter.prototype.triggerBuild = function () {
        this.emit(Action.TriggerBuild, { workerId: this.workerId });
    };
    MessageEmitter.prototype.error = function (error) {
        this.emit(Action.Error, { workerId: this.workerId, error: error });
    };
    MessageEmitter.prototype.warning = function (warning) {
        this.emit(Action.Warning, { workerId: this.workerId, warning: warning });
    };
    MessageEmitter.prototype.createTask = function (taskId) {
        this.emit(Action.CreateTask, { workerId: this.workerId, taskId: taskId });
    };
    MessageEmitter.prototype.startTask = function (taskId) {
        this.emit(Action.StartTask, { workerId: this.workerId, taskId: taskId });
    };
    MessageEmitter.prototype.succeedTask = function (taskId) {
        this.emit(Action.SucceedTask, { workerId: this.workerId, taskId: taskId });
    };
    MessageEmitter.prototype.failTask = function (taskId) {
        this.emit(Action.FailTask, { workerId: this.workerId, taskId: taskId });
    };
    MessageEmitter.prototype.skipTask = function (taskId) {
        this.emit(Action.SkipTask, { workerId: this.workerId, taskId: taskId });
    };
    return MessageEmitter;
}());
exports.MessageEmitter = MessageEmitter;
